### ISSUE TITLE
Recurring Transactions transfer date

### ISSUE THREAD

**Issue by Stonyjohn** [moneymanagerex/moneymanagerex#561]
MoneyManagerEx 1.2.2.
The date shown on the listing when a Recurring Transaction is transfered always reverts to the current date. Also there is general confusion between "Payment Date" and "Due Date".I have detailed the problem here:
http://forum.moneymanagerex.org/viewtopic.php?f=5&t=6045&start=10#p12687

**Comment by stef145g** [comment]
I will have time to re look at this issue in January 2016

**Comment by vomikan** [comment]
UP

**Comment by stef145g** [comment]
This issue has been revised as follows:
![recurringtransscreen](https://cloud.githubusercontent.com/assets/6927151/12698950/f09f5a44-c7fe-11e5-833d-a427d2e2f96d.png)

- Labels have been reclassified to better reflect the functionality
- The Date Reset button will now set the calendar and the Date Paid to today in edit mode, and only reset the Date Paid in transaction entry mode.
- On transaction entry, the calendar shows the Date Paid, and the entry date is set to the Date Due
- Automatic entries will use the Date Due as the transaction date.

Hopefully this fixes all issues that users have reported.

**Comment by Stonyjohn** [comment]
Thank you for working on this. I can now make manual enrties from the Recurring Transactions list without the transaction date reverting to the current date. However, I'm still confused over the difference between "Date Due" and "Date Paid".

As I explained before, I like to enter the "Next Occurrence" (right-click context menu) of items in the list 30 days before they are due, so I can plan my account usage over the next month. I need these transactions to be entered showing (on the account listing) the date these transactions apply, NOT the date the money is actually transferred. So there MUST be a difference between the "Date Paid" and the "Date Due". You state that automatic entries will use the Due Date as the transaction date. I take that to mean that the Due Date will be shown as the "Date Paid" on the account listing. So, to get that entry to be applied to the listing 30 days in advance, the "Date Paid" entry in the edit box must be 30 days before the "Date Due". I found that impossible to achieve.

In this context "Date Paid" must be the date the transaction is transferred from the "Recurring Transactions" listing to the "Account Listing".

**Comment by stef145g** [comment]
@Stonyjohn , I liked your initial request, but the only way to achieve an entry before the due date that would suite everybody was to split the dates. For my usage, I would not want my transactions entered 30 days before, but rather on the day I actually paid the bill. This was the reason for my initial strategy. Changing the strategy again hopefully will satisfy everyone's needs.

The current strategy is as follows:
**Old style setup**
- Use the Transaction Dates calendar to set both dates. This ensures that in either manual or automatic mode, the date requested by the user will be the date used in the actual transaction.

**New Style Setup**
- The "Transaction Dates" calendar  is used to set both dates.
- The "Date Paid" becomes the date when the automatic mode will react to that payment or the date I am deemed to make the payment.
- The "Date Due" would generally be the date when this bill is due. At some future date. This needs to be set after using the calendar control.
- In Semiautomatic mode the user is requested to enter this transaction on the "Date Paid".
- In fully automatic mode, the transaction is entered regardless of the user.
- On entry in any mode, the date of the transaction will be the "Date Due".
- On entry in manual or semiautomatic mode, the user can change the transaction date to today if necessary using the "Date Reset" button.
- On entry, the pending transaction will display the "Date Due", and the calendar will show the initial "Date Paid" being the requested date of user interaction.

The standard pre-notification of 14 days on the home page, is set by the "Date Paid".

@Stonyjohn , The individual dates need to be set after using the calendar control, to achieve what you want. Only the latest code will implement this new change and also to answer your last observations: you are correct.

**Comment by Stonyjohn** [comment]
Quote: "... the only way to achieve an entry before the due date that would suite everybody was to split the dates."
Surely, separating the dates is what you have done. On version 1.1.2 there is only one date column: "Next Due Date". On version 1.2.3 there are two: "Payment Date" and "Due Date". My policy is to make a payment on the due date - why pay early! However, at any time I want to pay early I simply register the transaction on the early date I actually pay - that date effectively becoming the due date. That would be the "Date Paid" on MMEX.

Once you have "Date Due" and "Date Paid" fields, these are essentially separate dates. The only omission is a date for action in MMEX when this "Recurring Transaction" is applied to the listing. This is indicated by right-clicking a transaction on the "Recurring Transactions" listing - the first item being "Enter next Occurrence...". It's that next occurrence date I need as the "secondl" date field.

To summarize:as far as MMEX is concerned the Date Due and Date Paid are the same (one database field only required), and the second date field is the date the transaction is entered onto the account listing. I have been doing that transfer manually (right-clicking the item), but now that you have added an automatic transfer facility, I was hoping my manual actions could now be automated, to occur when I want them to, whether on the due date or some time in advance.
.
Secondly, your statement:
 The standard pre-notification of 14 days on the home page, is set by the "Date Paid".
is a mystery. I see nothing resembling it on the home page. Please explain.

**Comment by stef145g** [comment]
What I meant by:
`The standard pre-notification of 14 days on the home page, is set by the "Date Paid".`
 is: 
Any transaction that is coming up within 14 days will be displayed on the Home Page under the heading **Upcoming Transactions**. The "Date Paid" field will now be the date which determines when this summary is displayed the home page. In version 1.2.0 this is known as the "Payment Date". Before version 1.2.0 this was "Next Occurrence".  The "Payment Date" has been renamed to "Date Paid".

With the introduction of the "Payment Date", the original "Next Occurrence" date became the "Due Date" in version 1.2.0. In the next major/minor release it will be known as "Date Due".

In Version 1.2.0 the payment date became today's date, which was not very popular and introduced some unexpected bugs. A reset button was introduced to set the transaction date to the "Due Date"

Future releases of 1.3 x will have the functionality as explained above. With the latest changes, the next occurrence will indeed be the "Date Due" which is the second date.

By splitting the dates, I can now see when to expect the bill, and record the actual date I pay the bill without affecting the future bill's expected due date. If I don't require the split, I can still set them with the same dates. 

The names of the fields becomes difficult, because we want to express the meaning in with small titles. If anyone has better names, they will be considered.

**Comment by Stonyjohn** [comment]
Thanks for the explanation.

So, the "Pay Date" item in the Edit Recurring Transaction box refers to the date appearing in the "Payment Date" column on the Recurring Transactions listing, and this is the date the home page notification appears. Is this the date the transaction would be transferred to the account listing if it has been set to "Automatic"? And am I right in thinking that the date of the transaction would appear on the actual account database is the date shown as the "Due Date" in Recurring Transactions"? This terminology is taken directly from the latest version - 1.2.3.

I need to get this right before my next occurrence showing in the "Payment" column in the Recurring Transactions listing, which is in 8 days time.

If I am right, you can see the confusion in the naming of the different items in the process, as they are not consistent, leading to confusion in how it all works.

If I'm completely wrong, I am confused anyway!

**Comment by stef145g** [comment]
I have set up 3 recurring transactions in a test database and created these images.
This is using version 1.2.5: New changes don't exist in this version. Some columns have been hidden.
![v1 2_recurring_trans](https://cloud.githubusercontent.com/assets/6927151/12867131/9a7ac10e-cd39-11e5-8eab-3c6722e6f6ef.png)

Today being the 7th February, you will notice that the home page only displays 2 transaction.
The one for today, and the one for the 21st February in 14 days time.
![v1 2_recurring_trans_home](https://cloud.githubusercontent.com/assets/6927151/12867164/81d1955a-cd3a-11e5-82d3-edcfe11cfb33.png)


The next picture is a combination of the Recurring Transactions, the Edit Recurring Transactions dialog, and the relationship that exists between them.
![v1 2_recurring_edit](https://cloud.githubusercontent.com/assets/6927151/12867414/12d978aa-cd41-11e5-93ee-5ea7adaff7d4.png)
- **Red** shows the Pay Date and Payment Date refer to the same thing.
In future, this label will become Date Paid

- **Blue** shows the Due date on both screens.
In future, this label will become Date Due

- **Green** shows the action of the date reset button. This date is replaced by the Due date.
In future, the  reset button will set this field to Today's date. and in transaction entry mode will contain the Date Due

@Stonyjohn  To answer your questions: To the first two questions, the answer is yes. To the third question, the answer is no for your current version but you can work around this by using the Date Reset Button. On entering a transaction, use the date reset button to correct this date. This will not work for you in the fully automated mode. In the semi automated mode, before confirming the entry, use the Date Reset button. This will reset the Payment Date to the Due Date.

In future the date shown on transaction entry and entered in the database will be set to the due date renamed "Date Due". This will also correct the fully automated entry by using the Date Due instead of the payment date. A display of the future screen is shown in a previous post above.

**Comment by stef145g** [comment]
@Stonyjohn  Correction... 4 questions not 3
You question: 
_Is this the date the transaction would be transferred to the account listing if it has been set to "Automatic?_
The answer is Yes, and No your next question. You would be presented with today's date in semiautomatic mode which you could correct as explained above, but the fully automatic mode transaction, would contain the payment date, which is the first date, and not the second date that you want. 

In future versions this is corrected as explained above.

**Comment by Stonyjohn** [comment]
Thanks for that - I think I've got it now.

I think it would be easier to understand if the date the next occurrence of transferring a transaction to the database was labelled "next occurrence" instead of Date Paid, as it wouldn't actually be paid on that date, but a date in the future.

I look forward to seeing the next version.

**Comment by Stonyjohn** [comment]
Following this correspondence back in February, I find that I still need 
to click the "Reset date" button to get the desired effect. Setting the 
transaction to "Automatic" stilL puts the transaction in showing the Due 
Date, when I need it to show the Payment Date.

**Comment by stef145g** [comment]
I don't quite understand what you are trying to achieve.

The logic behind this is as follows:
1. Having a single date, means that there is only 1 ability. That is, when to set a payment date.
2. Having 2 dates allows us to set a range of dates. This range is the possible dates that this bill is actually paid.

Having the start date, tells us when we need to activate this event.
Setting the start date to the same as the end date, gives us the same conditions as point 1.

If we want to set the payment date on the first of the month for the 20th of the month, we can do this. If we manually intervene, only we can actually determine when the payment is actually made. Otherwise we set the payment date to the due date. As you have stated before: "why would you want to pay the bill before it is due?"

So how can we tell the system that we have changed our mind and paid it today, rather when it was due? This is the function of the Reset Date button to help us determine today's date.

I don't know of any other possible way of setting this up. Perhaps you can determine the logic on how this functionality should work.
